src/tests/test_index_crash_recovery.cpp

```diff
@@ -60,6 +62,121 @@ void log_obj_life_counter() {
     LOGINFO("Object Life Counter\n:{}", str);
 }
 
+enum class OperationType {
+    Put,
+    Remove,
+};
+
+using Operation = std::pair< uint64_t, OperationType >;
+using OperationList = std::vector< Operation >;
+
+class SequenceGenerator {
+public:
+    SequenceGenerator(int putFreq, int removeFreq, uint64_t start_range, uint64_t end_range) :
+            putFreq_(putFreq), removeFreq_(removeFreq), start_range_(start_range), end_range_(end_range) {
+        std::random_device rd;
+        gen_ = std::mt19937(rd());
+        keyDist_ = std::uniform_int_distribution<>(start_range_, end_range_);
+        updateOperationTypeDistribution();
+    }
+
+    void setPutFrequency(int putFreq) {
+        putFreq_ = putFreq;
+        updateOperationTypeDistribution();
+    }
+
+    void setRemoveFrequency(int removeFreq) {
+        removeFreq_ = removeFreq;
+        updateOperationTypeDistribution();
+    }
+
+    void setRange(uint64_t start_range, uint64_t end_range) {
+        start_range_ = start_range;
+        end_range_ = end_range;
+        keyDist_ = std::uniform_int_distribution<>(start_range_, end_range_);
+    }
+
+    OperationList generateOperations(size_t numOperations, bool reset = false) {
+        std::vector< Operation > operations;
+        if (reset) { this->reset(); }
+        for (size_t i = 0; i < numOperations; ++i) {
+            uint32_t key = keyDist_(gen_);
+            auto [it, inserted] = keyStates.try_emplace(key, false);
+            auto& inUse = it->second;
+
+            OperationType operation = static_cast< OperationType >(opTypeDist_(gen_));
+
+            if (operation == OperationType::Put && !inUse) {
+                operations.emplace_back(key, OperationType::Put);
+                inUse = true;
+            } else if (operation == OperationType::Remove && inUse) {
+                operations.emplace_back(key, OperationType::Remove);
+                inUse = false;
+            }
+        }
```

Comment:
Better to use `operations.size()` instead of `i` to make sure we're returning a list with exactly the given size.